exporter/alibabacloudlogserviceexporter/logsdata_to_logservice.go

```diff
@@ -0,0 +1,205 @@
+	numDroppedLogs := 0
+	slsLogs := make([]*sls.Log, 0)
+	rls := ld.ResourceLogs()
+	for i := 0; i < rls.Len(); i++ {
+		rl := rls.At(i)
+		if rl.IsNil() {
+			continue
+		}
+
+		ills := rl.InstrumentationLibraryLogs()
+		for j := 0; j < ills.Len(); j++ {
+			ils := ills.At(j)
+			if ils.IsNil() {
+				continue
+			}
+
+			logs := ils.Logs()
+			for j := 0; j < logs.Len(); j++ {
+				lr := logs.At(j)
+				if lr.IsNil() {
+					continue
+				}
+				slsLog := mapLogRecordToLogService(lr, logger)
+				if slsLog == nil {
+					numDroppedLogs++
+				} else {
+					slsLogs = append(slsLogs, slsLog)
+				}
+			}
+		}
+	}
+
+	return slsLogs, numDroppedLogs
+}
+
+func mapLogRecordToLogService(lr pdata.LogRecord, logger *zap.Logger) *sls.Log {
+	if lr.Body().IsNil() {
+		return nil
+	}
+	var slsLog sls.Log
+
+	// pre alloc, refine if logContent's len > 16
+	slsLog.Contents = make([]*sls.LogContent, 0, 16)
+
+	slsLog.Contents = append(slsLog.Contents, &sls.LogContent{
+		Key:   proto.String(slsLogTimeUnixNano),
+		Value: proto.String(strconv.FormatUint(uint64(lr.Timestamp()), 10)),
+	})
+
+	slsLog.Contents = append(slsLog.Contents, &sls.LogContent{
+		Key:   proto.String(slsLogSeverityNumber),
+		Value: proto.String(strconv.FormatInt(int64(lr.SeverityNumber()), 10)),
+	})
+
+	slsLog.Contents = append(slsLog.Contents, &sls.LogContent{
+		Key:   proto.String(slsLogSeverityText),
+		Value: proto.String(lr.SeverityText()),
+	})
+
+	slsLog.Contents = append(slsLog.Contents, &sls.LogContent{
+		Key:   proto.String(slsLogName),
+		Value: proto.String(lr.Name()),
+	})
+
+	fields := map[string]interface{}{}
+	lr.Attributes().ForEach(func(k string, v pdata.AttributeValue) {
+		fields[k] = convertAttributeValue(v, logger)
+	})
+	attributeBuffer, _ := json.Marshal(fields)
+	slsLog.Contents = append(slsLog.Contents, &sls.LogContent{
+		Key:   proto.String(slsLogAttribute),
+		Value: proto.String(string(attributeBuffer)),
+	})
+
+	slsLog.Contents = append(slsLog.Contents, &sls.LogContent{
+		Key:   proto.String(slsLogContent),
+		Value: proto.String(convertAttributeValueToString(lr.Body(), logger)),
+	})
+
+	slsLog.Contents = append(slsLog.Contents, &sls.LogContent{
+		Key:   proto.String(slsLogFlags),
+		Value: proto.String(strconv.FormatUint(uint64(lr.Flags()), 16)),
+	})
+
+	slsLog.Contents = append(slsLog.Contents, &sls.LogContent{
+		Key:   proto.String(traceIDField),
+		Value: proto.String(lr.TraceID().HexString()),
+	})
+
+	slsLog.Contents = append(slsLog.Contents, &sls.LogContent{
+		Key:   proto.String(spanIDField),
+		Value: proto.String(lr.SpanID().HexString()),
+	})
+
+	if lr.Timestamp() > 0 {
+		// convert time nano to time seconds
+		slsLog.Time = proto.Uint32(uint32(lr.Timestamp() / 1000000000))
+	} else {
+		slsLog.Time = proto.Uint32(uint32(time.Now().Unix()))
+	}
+
+	return &slsLog
+}
+
+func convertAttributeValueToString(value pdata.AttributeValue, logger *zap.Logger) string {
```

Comment:
This appears very similar to https://github.com/open-telemetry/opentelemetry-collector/blob/8b67a09d745d9429a1d21f136d7f4283d56dd9be/translator/trace/protospan_translation.go#L114
Can you use that instead or there are differences?

Reply:
Great! `tracetranslator.AttributeValueToString` is what I want, Thanks a lot ~